src/algorithms/reco/TrackClusterMatch.cc

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2025 Tristan Protzman
+
+#include "algorithms/reco/TrackClusterMatch.h"
+#include "algorithms/reco/TrackClusterMatchConfig.h"
+#include <spdlog/logger.h>
+
+#include <edm4hep/Vector3f.h>
+#include <edm4hep/utils/vector_utils.h>
+
+#include <edm4eic/Cluster.h>
+#include <edm4eic/ClusterCollection.h>
+#include <edm4eic/Track.h>
+#include <edm4eic/TrackCollection.h>
+#include <edm4eic/TrackPoint.h>
+#include <edm4eic/TrackSegment.h>
+
+#include <DD4hep/Detector.h>
+
+namespace eicrecon {
+    void TrackClusterMatch::init(std::shared_ptr<spdlog::logger> logger, const dd4hep::Detector* detector) {
+        m_log = logger;
+        m_detector = detector;
+    }
+
+    void TrackClusterMatch::execute(const TrackClusterMatch::Input& input, const TrackClusterMatch::Output& output) const {
+        auto [tracks, clusters] = input;
+        auto [matched_particles] = output;
+        m_log->trace("We have {} tracks and {} clusters", tracks->size(), clusters->size());
+
+        std::set<int> used_tracks;
+        // Loop across each cluster, and find the cloeset projected track
+        for (auto cluster : *clusters) {
+            const double cluster_eta = edm4hep::utils::eta(cluster.getPosition());
+            const double cluster_phi = edm4hep::utils::angleAzimuthal(cluster.getPosition());
+            m_log->trace("Cluster at eta={}, phi={}", cluster_eta, cluster_phi);
+            // TODO: Get the detector ID so I can check if a projection is in the appropriate calorimeter
+
+            std::optional<edm4eic::TrackSegment> closest_segment;
+            int closest_segment_id = -1;
+            double closest_distance = m_cfg.matching_distance;
+            // Loop through each track segment, and its points
+            for (int closest_id = 0; auto track : *tracks) {
+                if (used_tracks.count(closest_id) > 0) {
+                    m_log->trace("Skipping track segment already used");
+                    continue;
+                }
+                for (auto point : track.getPoints()) {
+                    // Check if the point is at the calorimeter
+                    // int id = m_detector->volumeManager().lookupDetector(cluster.getHits()[0].getCellID()).id(); // TODO: Find programmatic way to get detector cluster is from
+                    uint32_t ecal_barrel_id = 101;
+                    uint32_t hcal_barrel_id = 111;
```

Comment:
No hardcoded numbers, please. Should be something like
`detector->constant<int>(m_cfg.hcal_id_constant);`
and
`ParameterRef<double> m_matching_distance {this, "hcalIdConstant", config().hcal_id_constant};`

Reply:
Is there a way to get the detector ID from the calorimeter cluster hit?  I experimented with `int id = m_detector>volumeManager().lookupDetector(cluster.getHits([0].getCellID()).id();`, but that felt a bit convoluted to me.

Reply:
`detector->constant` is the de-facto standard to do this. It will fetch values from the
```xml
<constant name="EcalBarrel_ID"        value="101"/>
<constant name="HcalBarrel_ID"      value="111"/>
```
from the compact files.